### PR TITLE
CLOUDP-32535 refactor and add tests

### PR DESCRIPTION
**Pull request by chunmingli**
https://jira.mongodb.org/browse/CLOUDP-32535

Divided into several more self contained files:
* stats.go (renamed from status.go) stats tracking
* utils.go walking directories, getting file list and calculating total size
* handlers.go HTTP route handlers
* config.go config parsing and validating

Tests added
* stats_test.go
* utils_test.go

To add (will be adding soon while other tests being reviewed)
* handlers_test.go

For config I'm a little torn about adding test coverage, but should be easy to add in general.

Added an Evergreen function `run tests` as well an ant target `ant test`.
Patch here:
https://evergreen.mongodb.com/build/atlasrestore_test_build_patch_b29925bc70af633f06d332ce801af0e7237fbf34_5b7369e42a60ed0c5f608b51_18_08_14_23_46_45
Or by yourself: 
`evergreen --config=/path/to/your/.evergreen.yml patch --project=atlasrestore --tasks=TEST_BUILD --variants=build_test --finalize`

Review: @lawctan 
CC: @Isabel11

### CODE REVIEW

File: stats.go [hunk 1]
```diff
@@ -46,66 +48,91 @@ type DownloadStats struct {
...
 	jobId      string
 	authKey    string
+	*time.Ticker
 }
 
-func NewStatsTracker(centralUrl, jobId, authKey string, stats *DownloadStats) *StatsTracker {
-	tracker := &StatsTracker{
-		stats:      stats,
+func (p *StatsTracker) Start() {
+	go func() {
```
Comment (lawctan) [inline stats.go]: either we add panic handler here, or file a separate ticket for it

File: stats.go [hunk 2]
```diff
@@ -46,66 +48,91 @@ type DownloadStats struct {
...
+	pw.updateRequestProgress(uint64(n))
+	return n, nil
 }
 
+type ReqMap map[uint64]*RequestStat
+
 type StatsTracker struct {
-	stats      *DownloadStats
+	stats *DownloadStats
+	*http.Client
```
Comment (lawctan) [inline stats.go]: not familiar with this syntax. the string after the last dot becomes the variable name?

File: stats_test.go [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+		var postedStats StatsToPost
+		json.Unmarshal(buf, &postedStats)
+
+		assertEqual(t, "Failed count mismatches", int32(1), postedStats.Failed)
+		assertEqual(t, "Succeeded count mismatches", int32(1), postedStats.Succeeded)
+		assertEqual(t, "InProgress count mismatches", int32(1), postedStats.InProgress)
+		assertEqual(t, "Array length mismatches", 3, len(postedStats.AllRequestStats))
+
+		w.Header().Set("Content-Type", "application/json")
+		fmt.Fprintln(w, `{"fake mms json response"}`)
```
Comment (lawctan) [inline stats_test.go]: fake mms json response is not valid json :) But if test passes it doesn't matter

File: stats_test.go [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+		ts.URL,         // centralUrl
+		"123abc",       // jobId
+		"IAmAnAuthKey", //authKey
+		stats,
+		time.NewTicker(time.Second*time.Duration(30)),
+		&http.Client{},
+	)
+
+	if err := tracker.Push(); err != nil {
+		t.Errorf("Posting failed")
```
Comment (lawctan) [inline stats_test.go]: let's pass error to ErrorF, or I saw another call above that just uses Error() and passes the err as argument

File: .evergreen.yml [hunk 5]
```diff
@@ -38,6 +38,14 @@ functions:
           echo "export GOROOT=/opt/golang/go1.9.5" >> activate-go
           echo "export PATH=$GOROOT/bin:$PATH" >> activate-go
 
+  "run tests":
+    - command: shell.exec
+      params:
+        working_dir: atlasrestore
+        script: |
+          source ./activate-go
```
[the remark this answers is not in the document]
Comment (chunmingli) [reply .evergreen.yml]: Generated here in the "golang version setup":
https://github.com/10gen/atlasrestore/blob/b29925bc70af633f06d332ce801af0e7237fbf34/.evergreen.yml#L37-L39

File: stats.go (hunk 2 above)
Comment (chunmingli) [reply stats.go]: Yep, unnamed fields are called "embedded field"s and the Type becomes the name when creating a struct literal. See
https://golang.org/ref/spec#Struct_types and
https://stackoverflow.com/questions/12537496/initialize-embedded-struct-in-go

File: stats_test.go (hunk 3 above)
Comment (chunmingli) [reply stats_test.go]: I'm gonna add another failure test case to this.

Comment (chunmingli) [reply stats_test.go]: The reason it passes is that when posting stats to MMS the golang server only cares about the response status code, not the content.

**Pull request merged**